core/cas-server-core-tickets/src/test/java/org/apereo/cas/ticket/registry/DefaultTicketRegistryTests.java

```diff
@@ -24,61 +46,148 @@
+            assertNull(reg.encodeTicket(null));
+            assertNotNull(reg.decodeTicket(mock(Ticket.class)));
+        }
+
+        @RepeatedTest(1)
+        void verifyGetExpiredTicketEventsSend() throws Throwable {
+            val applicationContext = mock(ConfigurableApplicationContext.class);
+            val registry = new DefaultTicketRegistry(mock(TicketSerializationManager.class), new DefaultTicketCatalog(),
+                    applicationContext);
+            val originalAuthn = CoreAuthenticationTestUtils.getAuthentication();
+            val ticketGrantingTicket = new TicketGrantingTicketImpl(TestTicketIdentifiers.generate().ticketGrantingTicketId(),
+                    originalAuthn, new HardTimeoutExpirationPolicy(1));
+            registry.addTicket(ticketGrantingTicket);
+            Thread.sleep(1500);
+            val tgtId = ticketGrantingTicket.getId();
+            val tgt = registry.getTicket(tgtId);
+            assertNull(tgt);
+            val internalTgt = registry.getMapInstance().get(tgtId);
+            assertNull(internalTgt);
+            verify(applicationContext).publishEvent(any(CasRequestSingleLogoutEvent.class));
+            verify(applicationContext).publishEvent(any(CasTicketGrantingTicketDestroyedEvent.class));
+        }
     }
 
-    @RepeatedTest(2)
-    void verifyCountForService() throws Throwable {
-        val service = RegisteredServiceTestUtils.getService(UUID.randomUUID().toString());
-        val registry = getNewTicketRegistry();
-        val user = UUID.randomUUID().toString();
+    @SpringBootTest(classes = LogoutTestConfiguration.class,
+        properties = {
+            "cas.ticket.tgt.core.only-track-most-recent-session=false",
+            "cas.ticket.registry.cleaner.schedule.enabled=false"
+        })
+    @ExtendWith(CasTestExtension.class)
+    @Nested
+    class LogoutTests {
 
-        for (int i = 0; i < 5; i++) {
-            val tgt = new MockTicketGrantingTicket(user);
-            val st = tgt.grantServiceTicket(service, TicketTrackingPolicy.noOp());
-            registry.addTicket(st);
-            registry.updateTicket(tgt);
+        @RepeatedTest(1)
+        void verifyGetExpiredTicketLogoutPerformed(final ConfigurableApplicationContext applicationContext) throws Throwable {
+            val registry = new DefaultTicketRegistry(mock(TicketSerializationManager.class), new DefaultTicketCatalog(),
+                    applicationContext);
+            val originalAuthn = CoreAuthenticationTestUtils.getAuthentication();
+            val ticketGrantingTicket = new TicketGrantingTicketImpl(BaseTicketRegistryTests.TestTicketIdentifiers.generate().ticketGrantingTicketId(),
+                    originalAuthn, new HardTimeoutExpirationPolicy(1));
+            registry.addTicket(ticketGrantingTicket);
+            Thread.sleep(1500);
+            val tgtId = ticketGrantingTicket.getId();
+            val tgt = registry.getTicket(tgtId);
```

Comment:
It looks like you might be able to await() here as well, [instead of a sleep](https://medium.com/cloud-native-daily/why-is-testing-with-thread-sleep-a-bad-idea-and-what-can-you-do-instead-3aad1d26a787)?

Reply:
Yes, `Thread.sleep` is redundant here. Will fix.

Reply:
In fact, the  `Thread.sleep` is necessary before the first checks: `val tgt = registry.getTicket(tgtId);     assertNull(tgt);`. I will keep it.